platforms/core-runtime/INSTRUMENTATION.md

```diff
@@ -0,0 +1,437 @@
+These inputs are often read via an API that Gradle doesn’t control, e.g. Java API via System.getenv().
+
+So to track these inputs we had to find a way to track these usages. 
+And that is why we started instrumenting bytecode.
+
+With a requirement to migrate Gradle core APIs to lazy properties, we were looking at how we could make old plugins compatible with new Gradle core APIs. 
+We decided that upgrading old compiled bytecode would work well. And reusing configuration cache bytecode instrumentation seemed like a good tool to reuse.
+
+So configuration cache instrumentation and API upgrades are using the same infrastructure now.
+
+
+## Functional design
+
+Imagine we have in Gradle core a task:
+```
+abstract class JavaCompile {
+    
+    private String sourceCompatibility = null;
+    
+    @Input
+    public String getSourceCompatibility() {
+        return sourceCompatibility;
+    }
+    
+    public void setSourceCompatibility(String sourceCompatibility) {
+        this.sourceCompatibility = sourceCompatibility;
+    }
+} 
+```
+and we want to change it to:
+
+```
+abstract class JavaCompile {
+    @Input
+    public Property<String> getSourceCompatibility();
+} 
+```
+
+To make old plugins compatible we need to replace all the usages of old JavaCompile with the new one.
+
+Let’s say it’s used in a plugin as:
+
+```
+fun applyPlugin(project: Project) {
+     project.tasks.named("compileJava", JavaCompile.class, task -> {
+	  task.setSourceCompatibility("1.8");
+     });
+}
+```
+
+We will then replace all such usages with bytecode instrumentation with:
+```
+fun applyPlugin(project: Project) {
+     project.tasks.named("compileJava", JavaCompile.class, task -> {
+	    JavaCompileInterceptorsDeclaration.intercept_setSourceCompatibility(task, "1.8");
+     });
+}
+```
+
+Where JavaCompileInterceptorsDeclaration is a class defined by Gradle developer to replace original logic and the replacement is done via bytecode instrumentation of plugin jars and dependencies.
+
+Due to that Instrumentation and bytecode upgrades is split into two big parts:
+1. Gradle distribution build time: Interceptor code generation and Gradle API type metadata collection
+2. Gradle runtime: Actual bytecode instrumentation
+
+JavaCompileInterceptorsDeclaration is defined by Gradle developer and with help of annotation processor Gradle then generates bytecode replacements. 
+The actual replacement happens when the user runs his Gradle build.
+
+**_Note: Groovy interception uses the same mechanism but replacements are a bit more complicated due to the dynamic nature of Groovy. 
+And due to that build logic is also intercepted when code is run and not just on the bytecode level._**
+
+Next sections will explain more in detail how we can define interceptors and how the code is then replaced.
+
+
+### Gradle Distribution Build Time
+
+At Gradle Distribution Build time we have two major parts:
+1. Declaring interceptors and generating interceptors
```

Comment:
❓ Declaring the interceptors happens before build time, i.e. when the developer adds the annotations to the source code, no? Shouldn't this say instead something like "we generate interceptors _from the annotations declared in the source code?_